### PR TITLE
Shadows uniform maps

### PR DESCRIPTION
**Pull request by bagnell**
Add per shadow map uniforms instead of a single automatic shadow map as a step toward having multiple shadow maps per scene.

### CODE REVIEW

File: Source/Scene/ShadowMap.js [hunk 1]
```diff
@@ -1248,6 +1258,59 @@ define([
         }
     };
 
+    ShadowMap.prototype.combineUniforms = function(uniforms, isTerrain) {
+        var isPointLight = this.isPointLight;
+        var bias = isPointLight ? this._pointBias : (isTerrain ? this._terrainBias : this._primitiveBias);
+
+        var that = this;
+        var mapUniforms = {
```
Comment (pjcozzi) [inline Source/Scene/ShadowMap.js]: Can any of these be packed into fewer uniforms, e.g., combine light direction/position?

File: Source/Shaders/Builtin/Functions/cascadeDistance.glsl [hunk 2]
```diff
@@ -1,4 +1,7 @@
+
+uniform vec4 u_shadowMapCascadeDistances;
```
Comment (pjcozzi) [inline Source/Shaders/Builtin/Functions/cascadeDistance.glsl]: Search *.glsl, but I _think_ we prefix uniforms like this with `czm_`.  Does it make sense in this context?

Comment (pjcozzi) [inline Source/Shaders/Builtin/Functions/cascadeDistance.glsl]: Same comment throughout.

**Comment by pjcozzi** [comment]
@lilleyse merge when you are happy with this.

File: Source/Shaders/Builtin/Functions/cascadeDistance.glsl (hunk 2 above)
Comment (bagnell) [inline Source/Shaders/Builtin/Functions/cascadeDistance.glsl]: We normally don't use uniforms like this. The `czm_` uniforms are set once per frame except `czm_model*`. The other option was passing them as arguments. The downside was that there was a lot of parameters for each `czm_` function. Perhaps put them all in a `struct` and use that as a parameter?

Comment (pjcozzi) [inline Source/Shaders/Builtin/Functions/cascadeDistance.glsl]: Ah, I think I have run into the same thing before and opted not to put a uniform with a function like this.  Pass them in.  You could use a `struct` if you trust GLSL compilers...could be OK.

File: Source/Scene/ShadowMap.js (hunk 1 above)
Comment (bagnell) [inline Source/Scene/ShadowMap.js]: I packed six of them together. More can be done with compression, but I would put that on the roadmap.

**Comment by bagnell** [comment]
@pjcozz This is ready for another look.

**Comment by pjcozzi** [comment]
:+1: 

@lilleyse merge when ready.

File: Source/Scene/ShadowMap.js [hunk 3]
```diff
@@ -1248,6 +1184,52 @@ define([
...
+            },
+            u_shadowMapTexelSizeDepthBiasAndNormalShadingSmooth : function() {
+                var texelStepSize = scratchTexelStepSize;
+                texelStepSize.x = 1.0 / that._textureSize.x;
+                texelStepSize.y = 1.0 / that._textureSize.y;
+
+                return new Cartesian4(texelStepSize.x, texelStepSize.y, bias.depthBias, bias.normalShadingSmooth);
+            },
+            u_shadowMapNormalOffsetScaleDistanceAndMaxDistance : function() {
+                return new Cartesian3(bias.normalOffsetScale, that._distance, that._maximumDistance);
```
Comment (lilleyse) [inline Source/Scene/ShadowMap.js]: We probably shouldn't be creating Cartesians all the time. Though this won't be a problem once all these uniforms are combined into structs for multiple shadow maps.

**Comment by lilleyse** [comment]
Looks good.

And just to summarize the changes I added
* The Shadows Sandcastle no longer does the hacky approach of updating the shaders because I couldn't get it to work with the uniform map changes here. Now the globe is just destroyed and recreated. The delay while waiting for the new globe to load isn't too bad.
* I removed a lot of the shadow map getters because they are no longer being called from `AutomaticUniforms`.

**Pull request merged**